function/app.py

```diff
@@ -0,0 +1,373 @@
+    members = response_members.json()['value']
+
+    # Fetch admins
+    url_admins = f"https://graph.microsoft.com/v1.0/groups/{group_id}/owners"
+    response_admins = requests.get(url_admins, headers=headers)
+    response_admins.raise_for_status()
+    admins = response_admins.json()['value']
+
+    # Combine members and admins
+    combined_members = members + admins
+
+    group_members_cache[group_id] = combined_members
+    return combined_members
+
+def get_identity_center_groups_and_relevant_users(identity_center_client, identity_store_id, group_name_prefix=""):
+    try:
+        logger.debug("Fetching AWS Identity Center groups and their memberships...")
+        groups = {}
+        relevant_users = set()
+
+        paginator = identity_center_client.get_paginator('list_groups')
+        for page in paginator.paginate(IdentityStoreId=identity_store_id):
+            for group in page['Groups']:
+                if group_name_prefix == "" or group['DisplayName'].startswith(group_name_prefix):
+                    groups[group['DisplayName']] = {
+                        'GroupId': group['GroupId'],
+                        'Members': set()
+                    }
+
+        for group_name, group_info in groups.items():
+            logger.info(f"Fetching members for group: {group_name}")
+            paginator = identity_center_client.get_paginator('list_group_memberships')
+            for page in paginator.paginate(IdentityStoreId=identity_store_id, GroupId=group_info['GroupId']):
+                for membership in page['GroupMemberships']:
+                    user_id = membership['MemberId']['UserId']
+                    username = get_identity_center_username(identity_center_client, identity_store_id, user_id)
+                    if username:
+                        group_info['Members'].add(username)
+                        relevant_users.add(user_id)
+
+        logger.info(f"Number of Identity Center groups retrieved with prefix '{group_name_prefix}': {len(groups)}")
+        return groups, relevant_users
+    except ClientError as e:
+        logger.error(f"Error listing Identity Center groups and memberships: {e}")
+        raise e
+
+def get_identity_center_username(identity_center_client, identity_store_id, user_id):
+    if user_id in user_cache:
+        return user_cache[user_id]
+
+    try:
+        logger.debug(f"Fetching username for user ID: {user_id}")
+        response = identity_center_client.describe_user(
+            IdentityStoreId=identity_store_id,
+            UserId=user_id
+        )
+        username = response.get('UserName')
+        if username:
+            user_cache[user_id] = username
+        return username
+    except ClientError as e:
+        logger.error(f"Error getting username for user ID {user_id}: {e}")
+        return None
+
+def get_identity_center_user_id_by_username(identity_center_client, identity_store_id, username):
+    if username in user_cache:
```

Comment:
Is there the potential to instead go something like:

```
for id, name in user_cache.items():
     if username == name:
          return id
```
Then add the user to the cache via the user_id rather than name on line 150. 

The reason I suggest this is the function above `get_identity_center_username` should in theory already have stored that information in the cache (I think, unless I'm misunderstanding what's returned). That way, you may make fewer (if any) API calls.

Reply:
I think you might be right -- there is at least one API method called twice when it needs to be called once. I might leave this for version 2 though I'm kinda scared to unpick it. :/